### ISSUE TITLE
react/no-danger fails to detect with styled components

### ISSUE THREAD

**Issue by RichardCzechowski** [yannickcr/eslint-plugin-react#2958]
It seems like react/no-danger cannot find instances of dangerouslySetInnerHTML within styled components.

```
// Test.tsx

import React from 'react'
import styled from 'styled-components'

const StyledDiv = styled.div``

export const Foo = ()=>{
    return <div
        dangerouslySetInnerHTML={{__html: 'foo'}}  // This will be detected
    ></div>
}
export const Bar = ()=>{
    return <StyledDiv
        dangerouslySetInnerHTML={{__html: 'foo'}}  // This will not be detected
    ></StyledDiv>
}
```

I've created a sample app here to illustrate. Just open App.tsx and change which component declaration is active.
https://github.com/RichardCzechowski/react-no-danger-bug

**Comment by ljharb** [comment]
Styled components are a template literal, and aren’t actually react components. It wouldn’t make sense for this plugin to start trying to dig into a template literal DSL, when jsx exists as the canonical syntactic alternative to React.createElement. You’d need a styled-components-specific eslint plugin for that.

Separately, “foo” and “bar” are lowercase, which means they’re not components. Components must start with a capital letter.

However, i think the real issue here is that dangerouslySetInnerHTML is just a random property on a custom component, which StyledDiv is. Only html components have that as a magic property, and that means, native non-styled-component ones.

**Comment by jbmanning** [comment]
I'd make the argument even if the component isn't a native DOM element, we could assume this specific property should be reported regardless. Or at least included in a config flag if you should check non-native elements.

**Comment by ljharb** [comment]
That argument does not hold. A custom component can choose to have a prop named `dangerouslySetInnerHTML` at zero risk - it's just unlikely.

**Comment by jbmanning** [comment]
I agree, it is possible for it to be a safe prop. That being said, it seems to be extra-ordinarily unlikely that a prop named `dangerouslySetInnerHTML` would not be "dangerous" to use.

**Comment by ljharb** [comment]
How do you envision that working?

**Comment by jbmanning** [comment]
Don't know the correct config syntax, but based on the current code I would envision something like this:

```typescript
if ((config.checkNonNativeNodes || jsxUtil.isDOMComponent(node.parent)) && isDangerous(node.name.name)) {
    context.report({
        ...
    });
}
```

**Comment by ljharb** [comment]
So you're envisioning an option that blindly bans the prop name on all components?

**Comment by jbmanning** [comment]
"blindly" sounds harsh, but yes. It would ban any props with dangerous names.

**Comment by ljharb** [comment]
Unless there's a styled-components parser, there'd be no way we could warn on things inside a template literal - and *with* such a parser, if it produced the same node names as any of the major existing parsers (espree, babel, typescript) then it would Just Work.

I don't think it would be appropriate to ban all components from using `dangerouslySetInnerHTML` - and you can do that yourself using `forbid-component-props`.

So, I'm going to close this.
